File: frontend/src/api/modeling.ts
import api from "./client";

export interface ModelTrainConfig {
  target: string;
  problem_type: string;
  model_name: string;
  config: {
    test_size: number;
    random_seed: number;
    class_weight?: string | null;
    cv_folds?: number;
    use_smote?: boolean;
    n_estimators?: number | null;
    max_depth?: number | null;
    learning_rate?: number | null;
    degree?: number | null;
    k?: number | null;
    kernel?: string | null;
    C?: number | null;
    apply_scaling?: boolean;
  };
  feature_selection: {
    method: string;
    selected_features?: string[] | null;
    max_features?: number | null;
  };
  feature_engineering?: {
    enabled: boolean;
    log_transform?: string[] | null;
    bin_columns?: string[] | null;
    interaction_pairs?: string[][] | null;
  };
}

export async function trainModel(
  file: Blob,
  filename: string,
  config: ModelTrainConfig
): Promise<any> {
  const formData = new FormData();
  formData.append("file", file, filename);
  formData.append("config_json", JSON.stringify(config));

  try {
    const response = await api.post("/models/train", formData, {
      headers: { "Content-Type": "multipart/form-data" },
    });
    return response.data;
  } catch (err: any) {
    // Extract backend error detail from 400/422 responses
    const detail = err?.response?.data?.detail;
    if (detail) {
      throw new Error(typeof detail === "string" ? detail : JSON.stringify(detail));
    }
    throw err;
  }
}

File: backend/app/schemas/modeling.py
from __future__ import annotations

from pydantic import BaseModel


class FeatureEngineeringConfig(BaseModel):
    log_transform: list[str] | None = None        # columns to log-transform
    bin_columns: list[str] | None = None           # columns to bin into quantiles
    interaction_pairs: list[list[str]] | None = None  # pairs of columns to multiply
    enabled: bool = False


class ModelConfig(BaseModel):
    test_size: float = 0.2
    random_seed: int = 42
    sample_size: int | None = None               # Max rows to use (None = all)
    class_weight: str | None = None  # "none" | "balanced"
    cv_folds: int = 5                             # k-fold cross-validation
    use_smote: bool = False                       # SMOTE for class imbalance
    # Model-specific hyperparams
    n_estimators: int | None = None       # RF, GB
    max_depth: int | None = None          # DT, RF
    learning_rate: float | None = None    # GB
    degree: int | None = None             # Polynomial
    k: int | None = None                  # KNN
    kernel: str | None = None             # SVM: "linear" | "rbf"
    C: float | None = None               # SVM
    apply_scaling: bool = False           # Auto-scale for KNN/SVM


class FeatureSelectionConfig(BaseModel):
    method: str = "none"  # "none" | "manual" | "forward" | "backward" | "lasso"
    selected_features: list[str] | None = None  # For manual selection
    max_features: int | None = None  # For forward/backward


class ModelTrainRequest(BaseModel):
    target: str
    problem_type: str  # "auto" | "classification" | "regression"
    model_name: str
    config: ModelConfig = ModelConfig()
    feature_selection: FeatureSelectionConfig = FeatureSelectionConfig()
    feature_engineering: FeatureEngineeringConfig = FeatureEngineeringConfig()


class CVMetrics(BaseModel):
    metric_name: str
    mean: float
    std: float
    folds: list[float]


class ClassificationMetrics(BaseModel):
    accuracy: float
    precision: float
    recall: float
    f1: float
    roc_auc: float | None = None
    confusion_matrix: list[list[int]]
    labels: list[str]
    cv: list[CVMetrics] | None = None


class RegressionMetrics(BaseModel):
    mae: float
    rmse: float
    r2: float
    cv: list[CVMetrics] | None = None


class FeatureImportance(BaseModel):
    feature: str
    importance: float


class ModelTrainResponse(BaseModel):
    model_name: str
    problem_type: str
    target: str
    features_used: list[str]
    classification_metrics: ClassificationMetrics | None = None
    regression_metrics: RegressionMetrics | None = None
    feature_importances: list[FeatureImportance] | None = None
    feature_engineering_applied: list[str] | None = None  # descriptions of FE steps
    smote_applied: bool = False
    # Chart data
    confusion_matrix_data: dict | None = None       # {z, x, y}
    roc_curve_data: dict | None = None              # {fpr, tpr, auc}
    predicted_vs_actual: dict | None = None         # {y_true, y_pred}
    residuals_data: dict | None = None              # {y_pred, residuals}
    coefficients: list[dict] | None = None          # [{feature, coef}]
